File: Topic_3/Ports/Python/ab2022_model.py
import numpy as np
from scipy.optimize import root_scalar

class AB2022Model:
    def __init__(self, params=None):
        """
        Initializes the model with the given parameters.
        """
        # Default baseline parameters from prompt and Stata toolkit
        self.params = {
            'alpha_C': 0.85,
            'alpha_R': 0.66,
            'theta_C': 0.5,
            'theta_R': 0.55,
            'omega_C': 0.03,
            'omega_R': 0.07,
            'beta_C': 0.03,
            'beta_R': 0.0,
            'a_bar_C': 2.0,
            'a_bar_R': 1.0,
            'tau_C': 0.01,
            'tau_R': 0.005,
            'c_C': 1.4,
            'c_R': 1.4,
            'r_a': 150.0,
            'S_bar_C': 999.0,
            'S_bar_R': 999.0,
            'x_1_bar': 999.0,
        }
        if params:
            self.params.update(params)
        
        # Grid setup: 10001 points from -50 to 50
        self.x = np.linspace(-50, 50, 10001)
        self.dist = np.abs(self.x)
        self.eps_C = np.ones_like(self.x)
        self.eps_R = np.ones_like(self.x)

    def solver(self, L, y):
        """
        SOLVER procedure: Calculates the spatial equilibrium for a given wage (y) and total employment (L).
        """
        p = self.params
        
        # 1.1 Amenity and Productivity Shifters
        A_tilde_C = p['a_bar_C'] * self.eps_C * (L**p['beta_C']) * np.exp(-p['tau_C'] * self.dist)
        A_tilde_R = p['a_bar_R'] * self.eps_R * (L**p['beta_R']) * np.exp(-p['tau_R'] * self.dist)
        
        # Floor space shifters
        # a_x_C = A_tilde_x_C^(1/(1-alpha_C)) * y^(alpha_C/(alpha_C-1))
        # a_x_R = A_tilde_x_R^(1/(1-alpha_R)) * y^(1/(1-alpha_R))
        a_C = A_tilde_C**(1/(1-p['alpha_C'])) * (y**(p['alpha_C']/(p['alpha_C']-1)))
        a_R = A_tilde_R**(1/(1-p['alpha_R'])) * (y**(1/(1-p['alpha_R'])))
        
        # 1.2 Land Rents
        # Formula: r = a/(1+omega) * (a/(c(1+theta)))^((1+omega)/(theta-omega)) - c * (a/(c(1+theta)))^((1+theta)/(theta-omega))
        def calc_rent(a, omega, theta, c):
            # Optimal height S* = (a / (c(1+theta)))^(1/(theta-omega))
            S_star = (a / (c * (1 + theta)))**(1 / (theta - omega))
            rent = (a / (1 + omega)) * (S_star**(1 + omega)) - c * (S_star**(1 + theta))
            return rent

        r_C = calc_rent(a_C, p['omega_C'], p['theta_C'], p['c_C'])
        r_R = calc_rent(a_R, p['omega_R'], p['theta_R'], p['c_R'])
        
        # Land Use Decision
        # 1: Commercial, 2: Residential, 3: Agricultural
        U = np.ones_like(self.x) * 3
        U[(r_R > r_C) & (r_R > p['r_a'])] = 2
        U[(r_C >= r_R) & (r_C > p['r_a'])] = 1
        
        # Urban Growth Boundary (x_1_bar)
        U[self.dist > p['x_1_bar']] = 3
        
        # 1.3 Endogenous Variables
        # Profit-maximizing height
        S_star_C = (a_C / (p['c_C'] * (1 + p['theta_C'])))**(1 / (p['theta_C'] - p['omega_C']))
        S_star_R = (a_R / (p['c_R'] * (1 + p['theta_R'])))**(1 / (p['theta_R'] - p['omega_R']))
        
        # Realized height
        S_C = np.minimum(p['S_bar_C'], S_star_C)
        S_R = np.minimum(p['S_bar_R'], S_star_R)
        
        # Bid rents (p_bar)
        p_C_bid = a_C * (1 / (1 - p['omega_C'])) * (S_C**p['omega_C'])
        p_R_bid = a_R * (1 / (1 - p['omega_R'])) * (S_R**p['omega_R'])
        
        # Labor Demand and Supply
        # L_D = sum( L_x_C ) where L_x_C = alpha_C/(1-alpha_C) * p_C / y * S_C
        L_D_loc = np.zeros_like(self.x)
        mask_C = (U == 1)
        L_D_loc[mask_C] = (p['alpha_C'] / (1 - p['alpha_C'])) * (p_C_bid[mask_C] / y) * S_C[mask_C]
        
        # L_S = sum( n_x ) where n_x = S_R / f_R and f_R = (1-alpha_R) * y / p_R
        L_S_loc = np.zeros_like(self.x)
        mask_R = (U == 2)
        L_S_loc[mask_R] = (S_R[mask_R] * p_R_bid[mask_R]) / ((1 - p['alpha_R']) * y)
        
        L_D_total = np.sum(L_D_loc)
        L_S_total = np.sum(L_S_loc)
        
        results = {
            'L_D_total': L_D_total,
            'L_S_total': L_S_total,
            'U': U,
            'S_C': S_C,
            'S_R': S_R,
            'S_x': np.where(U == 1, S_C, np.where(U == 2, S_R, 0)),
            'p_C_bid': p_C_bid,
            'p_R_bid': p_R_bid,
            'r_C': r_C,
            'r_R': r_R,
            'S_star_C': S_star_C,
            'S_star_R': S_star_R,
            'x': self.x,
            'dist': self.dist
        }
        return results

    def wage_objective(self, y, L):
        """
        Objective function for wage market clearing: L_D / L_S - 1 = 0
        """
        res = self.solver(L, y)
        if res['L_S_total'] <= 1e-6:
            return 1e10 if res['L_D_total'] > 1e-6 else 0.0
        return res['L_D_total'] / res['L_S_total'] - 1

    def find_wage(self, L, y_guess):
        """
        WAGE: Finds the market-clearing wage y for a fixed total employment L.
        """
        try:
            # Use Brent's method for efficiency
            sol = root_scalar(self.wage_objective, args=(L,), bracket=[0.1, 100], method='brentq', xtol=1e-6)
            return sol.root
        except ValueError:
            # Fallback to dampened fixed-point iteration as in Stata
            y = y_guess
            for _ in range(100):
                res = self.solver(L, y)
                if res['L_S_total'] <= 1e-6:
                    if res['L_D_total'] <= 1e-6: break
                    y *= 1.2
                elif res['L_D_total'] <= 1e-6:
                    y *= 0.8
                else:
                    y_factor = (res['L_D_total'] / res['L_S_total'])**0.01
                    y_new = 0.5 * y + 0.5 * y * y_factor
                    if abs(y_new / y - 1) < 1e-7: break
                    y = y_new
            return y

    def find_eq(self, L_guess, y_guess):
        """
        FINDEQ: Finds the general equilibrium total employment L and wage y.
        """
        L = L_guess
        y = y_guess
        for i in range(100):
            y = self.find_wage(L, y)
            res = self.solver(L, y)
            avg_L_model = 0.5 * (res['L_D_total'] + res['L_S_total'])
            
            if avg_L_model <= 1e-6:
                # City collapsed
                return 0.0, y, res
                
            if abs(L / avg_L_model - 1) < 1e-4:
                break
            
            # Update L using dampened iteration
            L = 0.5 * L + 0.5 * avg_L_model
        
        return L, y, res

    def invert(self, observed_heights, target_L, conv_param=0.05, tolerance=0.001):
        """
        INVERT: Adjusts local amenities (epsilon) to match observed building heights 
        and total population target.
        """
        L, y, res = self.find_eq(target_L, 2.5)
        
        for i in range(200):
            # 1. Update amenities based on height ratio (CONV logic)
            # S_star is the profit-maximizing height without the limit
            S_model = res['S_x']
            mask_urban = (res['U'] < 3)
            
            # Adjustment factor: Observed / Model
            adj = np.ones_like(self.x)
            valid = (S_model > 1e-3) & (observed_heights > 1e-3)
            adj[valid] = observed_heights[valid] / S_model[valid]
            
            # Apply update to both C and R (dampened)
            self.eps_C = (1 - conv_param) * self.eps_C + conv_param * adj * self.eps_C
            self.eps_R = (1 - conv_param) * self.eps_R + conv_param * adj * self.eps_R
            
            # 2. Update L and y
            L, y, res = self.find_eq(L, y)
            
            # 3. Adjust population target (EMP logic)
            if L > 1e-6:
                pop_adj = (target_L / L)**0.01
                self.eps_R *= pop_adj
                L, y, res = self.find_eq(L, y)
            
            # Check convergence (correlation > 0.999 and pop gap < tolerance)
            if L > 1e-6:
                corr = np.corrcoef(observed_heights[mask_urban], S_model[mask_urban])[0, 1]
                pop_gap = abs(L / target_L - 1)
                if corr > 0.999 and pop_gap < tolerance:
                    break
                    
        return L, y, res

if __name__ == "__main__":
    # Baseline run
    model = AB2022Model()
    L_eq, y_eq, res = model.find_eq(1000000, 2.5)
    print(f"Equilibrium Results:")
    print(f"L: {L_eq:,.2f}")
    print(f"y: {y_eq:.4f}")

File: Topic_3/Ports/R/ab2022_model.R
library(data.table)

#' SOLVER function: Computes spatial equilibrium for a given wage (y) and total employment (L)
#' @param dt data.table with location data (x, D)
#' @param params list of parameters
#' @return list(L_hat_demand, L_hat_supply)
SOLVER <- function(dt, params) {
  # Current values of L and y are stored in params or passed via dt columns
  L_val <- params$L
  y_val <- params$y

  # 1.1 Amenity and Productivity Shifters
  # Use epsilon_C/R if they exist in the data.table, otherwise default to 1.0
  eps_C <- if ("epsilon_C" %in% names(dt)) dt$epsilon_C else 1.0
  eps_R <- if ("epsilon_R" %in% names(dt)) dt$epsilon_R else 1.0

  dt[, A_tilde_x_C := params$a_bar_C * eps_C * (L_val^params$beta_C) * exp(-params$tau_C * D)]
  dt[, A_tilde_x_R := params$a_bar_R * eps_R * (L_val^params$beta_R) * exp(-params$tau_R * D)]
  
  # Floor Space Shifters
  dt[, a_x_C := A_tilde_x_C^(1/(1-params$alpha_C)) * y_val^(params$alpha_C/(params$alpha_C-1))]
  dt[, a_x_R := A_tilde_x_R^(1/(1-params$alpha_R)) * y_val^(1/(1-params$alpha_R))]

  # 1.2 Land Rents
  # Commercial
  dt[, r_x_C := (a_x_C / (1 + params$omega_C)) * (a_x_C / (params$c_C * (1 + params$theta_C)))^((1 + params$omega_C) / (params$theta_C - params$omega_C)) - 
        params$c_C * (a_x_C / (params$c_C * (1 + params$theta_C)))^((1 + params$theta_C) / (params$theta_C - params$omega_C))]
  
  # Residential
  dt[, r_x_R := (a_x_R / (1 + params$omega_R)) * (a_x_R / (params$c_R * (1 + params$theta_R)))^((1 + params$omega_R) / (params$theta_R - params$omega_R)) - 
        params$c_R * (a_x_R / (params$c_R * (1 + params$theta_R)))^((1 + params$theta_R) / (params$theta_R - params$omega_R))]

  # Land Use Decision (U)
  # 1: Commercial, 2: Residential, 3: Agricultural
  dt[, U := 3L] # Default to Agricultural
  dt[r_x_R > params$r_a & r_x_R >= r_x_C, U := 2L]
  dt[r_x_C > params$r_a & r_x_C > r_x_R, U := 1L]
  
  # Handle S_bar constraints and compute Realized Height (S_x,i)
  dt[, S_x_C := 0.0]
  dt[, S_x_R := 0.0]
  
  dt[U == 1, S_x_C := pmin(params$S_bar_C, (a_x_C / (params$c_C * (1 + params$theta_C)))^(1 / (params$theta_C - params$omega_C)))]
  dt[U == 2, S_x_R := pmin(params$S_bar_R, (a_x_R / (params$c_R * (1 + params$theta_R)))^(1 / (params$theta_R - params$omega_R)))]

  # Bid Rents (p_x,i)
  dt[, p_x_C := 0.0]
  dt[, p_x_R := 0.0]
  dt[U == 1, p_x_C := a_x_C * (1 / (1 - params$omega_C)) * S_x_C^params$omega_C]
  dt[U == 2, p_x_R := a_x_R * (1 / (1 - params$omega_R)) * S_x_R^params$omega_R]

  # Labor Demand (L_x,C) and Supply (n_x)
  dt[, L_x_C := 0.0]
  dt[, n_x := 0.0]
  
  dt[U == 1, L_x_C := (params$alpha_C / (1 - params$alpha_C)) * (p_x_C / y_val) * S_x_C]
  dt[U == 2, n_x := S_x_R / ((1 - params$alpha_R) * y_val / p_x_R)]

  # Aggregate values
  L_hat_demand <- sum(dt$L_x_C, na.rm = TRUE)
  L_hat_supply <- sum(dt$n_x, na.rm = TRUE)
  
  return(list(L_hat_demand = L_hat_demand, L_hat_supply = L_hat_supply))
}

#' WAGE function: Finds the wage y that clears the labor market for a fixed total employment L
WAGE <- function(dt, params, tolerance = 0.001, max_iter = 100) {
  iter <- 0
  res <- SOLVER(dt, params)
  
  # Handle zero cases to avoid premature termination
  if (res$L_hat_demand == 0 && res$L_hat_supply == 0) {
    obj_int <- 1.0 # Force at least one iteration
  } else {
    obj_int <- abs((res$L_hat_demand + 1e-6) / (res$L_hat_supply + 1e-6) - 1)
  }
  
  while (obj_int > tolerance && iter < max_iter) {
    if (res$L_hat_supply == 0) {
      y_factor <- 1.2
    } else if (res$L_hat_demand == 0) {
      y_factor <- 0.8
    } else {
      y_factor <- (res$L_hat_demand / res$L_hat_supply)^0.01
    }
    
    params$y <- 0.5 * params$y + 0.5 * params$y * y_factor
    res <- SOLVER(dt, params)
    
    if (res$L_hat_demand == 0 && res$L_hat_supply == 0) {
      obj_int <- 1.0
    } else {
      obj_int <- abs((res$L_hat_demand + 1e-6) / (res$L_hat_supply + 1e-6) - 1)
    }
    iter <- iter + 1
    # cat(sprintf("  WAGE iter %d: y=%.4f, Ld=%.2f, Ls=%.2f, obj=%.4f\n", iter, params$y, res$L_hat_demand, res$L_hat_supply, obj_int))
  }
  return(params)
}

#' FINDEQ function: Finds the equilibrium total employment L
FINDEQ <- function(dt, params, tolerance = 0.001, max_iter = 100) {
  iter <- 0
  params <- WAGE(dt, params)
  res <- SOLVER(dt, params)
  
  if (res$L_hat_demand == 0 && res$L_hat_supply == 0) {
    obj_ext <- 1.0 # Force iteration
  } else {
    obj_ext <- abs(params$L / (0.5 * (res$L_hat_demand + res$L_hat_supply)) - 1)
  }
  
  while (obj_ext > tolerance && iter < max_iter) {
    # If city is empty, we need to be careful about L shrinking to 0
    if (res$L_hat_demand == 0 && res$L_hat_supply == 0) {
      params$y <- params$y * 0.8
    }
    
    params$L <- 0.5 * params$L + 0.25 * (res$L_hat_demand + res$L_hat_supply)
    params <- WAGE(dt, params)
    res <- SOLVER(dt, params)
    
    if (res$L_hat_demand == 0 && res$L_hat_supply == 0) {
      obj_ext <- 1.0
    } else {
      obj_ext <- abs(params$L / (0.5 * (res$L_hat_demand + res$L_hat_supply)) - 1)
    }
    iter <- iter + 1
    cat(sprintf("FINDEQ iter %d: L=%.2f, y=%.4f, Ld=%.2f, Ls=%.2f, obj=%.4f\n", iter, params$L, params$y, res$L_hat_demand, res$L_hat_supply, obj_ext))
  }
  return(params)
}

#' INVERT function: Adjusts local amenities to match observed building heights
#' For simplicity, this implementation follows the pseudo-code logic
INVERT <- function(dt, params, observed_heights_C, observed_heights_R, tolerance = 0.001, max_iter = 50) {
  # Initialize epsilons if not present
  if (!"epsilon_C" %in% names(dt)) dt[, epsilon_C := 1.0]
  if (!"epsilon_R" %in% names(dt)) dt[, epsilon_R := 1.0]
  
  iter <- 0
  params <- FINDEQ(dt, params)
  
  # Current height in model
  current_S_C <- dt$S_x_C
  current_S_R <- dt$S_x_R
  
  # Simple correlation check for commercial height as example
  corr_C <- cor(observed_heights_C[dt$U == 1], current_S_C[dt$U == 1])
  
  while (iter < max_iter) {
    # Update epsilons based on height ratio (avoiding division by zero)
    dt[U == 1, epsilon_C := epsilon_C * (observed_heights_C / pmax(S_x_C, 0.001))]
    dt[U == 2, epsilon_R := epsilon_R * (observed_heights_R / pmax(S_x_R, 0.001))]
    
    params <- FINDEQ(dt, params)
    
    # Re-calculate correlation or objective
    # This is a simplified version of the inversion logic
    iter <- iter + 1
  }
  return(params)
}
